### PR TITLE
Adding GuestAccelerators to InstanceTemplate

### PR DESCRIPTION
**Pull request by jonasasx**
Resolves https://github.com/kubernetes/kops/issues/6517
Fix non-master PR https://github.com/kubernetes/kops/pull/13704

### CODE REVIEW

File: pkg/apis/kops/instancegroup.go [hunk 1]
```diff
@@ -363,3 +365,9 @@ type LoadBalancer struct {
 	// TargetGroupARN to associate with this instance group (AWS ALB/NLB)
 	TargetGroupARN *string `json:"targetGroupARN,omitempty"`
 }
+
+// AcceleratorConfig defines an accelerator config
+type AcceleratorConfig struct {
+	AcceleratorCount int64  `json:"acceleratorCount,omitempty"`
+	AcceleratorType  string `json:"acceleratorType,omitempty"`
```
Comment (hakman) [inline pkg/apis/kops/instancegroup.go:372]: Why not just `count` and `type`? Should be pretty obvious they refer to accelerators.
```suggestion
	Count int64  `json:"count,omitempty"`
	Type  string `json:"type,omitempty"`
```

File: pkg/model/gcemodel/autoscalinggroup.go [hunk 2]
```diff
@@ -90,7 +90,7 @@ func (b *AutoscalingGroupModelBuilder) buildInstanceTemplate(c *fi.ModelBuilderC
 				BootDiskImage:  s(ig.Spec.Image),
 
 				// TODO: Support preemptible nodes?
-				Preemptible: fi.Bool(false),
+				Preemptible: fi.Bool(ig.Spec.GuestAccelerators != nil && len(ig.Spec.GuestAccelerators) > 0),
```
Comment (hakman) [inline pkg/model/gcemodel/autoscalinggroup.go:93]: Why is this related to preemption?

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go [hunk 3]
```diff
@@ -211,6 +213,16 @@ func (e *InstanceTemplate) Find(c *fi.Context) (*InstanceTemplate, error) {
 		// System fields
 		actual.Lifecycle = e.Lifecycle
 
+		if p.GuestAccelerators != nil {
```
Comment (hakman) [inline upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:216]:
```suggestion
```

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go [hunk 4]
```diff
@@ -335,13 +347,26 @@ func (e *InstanceTemplate) mapToGCE(project string, region string) (*compute.Ins
 		})
 	}
 
+	var accelerators []*compute.AcceleratorConfig
+	if e.GuestAccelerators != nil {
```
Comment (hakman) [inline upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:351]:
```suggestion
```

File: pkg/model/gcemodel/autoscalinggroup.go [hunk 5]
```diff
@@ -168,6 +168,16 @@ func (b *AutoscalingGroupModelBuilder) buildInstanceTemplate(c *fi.ModelBuilderC
 			//}
 			//t.Labels = labels
 
+			if ig.Spec.GuestAccelerators != nil && len(ig.Spec.GuestAccelerators) > 0 {
```
Comment (hakman) [inline pkg/model/gcemodel/autoscalinggroup.go:171]: This `if` can be removed. Setting this to empty slice allows unsetting it later if not wanted anymore. Otherwise, a `nil` value will be ignored.
```suggestion
```

**Review by hakman — CHANGES_REQUESTED** [review]
Thanks @jonasasx. This looks mostly good, just some nits and Qs from my side.
Besides this there are a few things missing:
1. validation that both `count` and `type` are set.
2. add the fields to the terraform output also.

File: pkg/apis/kops/instancegroup.go (hunk 1 above)
Comment (jonasasx) [reply pkg/apis/kops/instancegroup.go:372]: gcp fields of InstanceTemplate are named with the prefix `accelerator`. I think we need to keep uniformity.

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 2 above)
Comment (jonasasx) [reply pkg/model/gcemodel/autoscalinggroup.go:93]: Google Cloud forbids to create instances with GPUs in `MIGRATE` maintenance mode. Only `TERMINATE` is allowed.

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 5 above)
Comment (jonasasx) [reply pkg/model/gcemodel/autoscalinggroup.go:171]: I have been coded in Java for many years with endless NullPointerExceptions :)

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go (hunk 3 above)
Comment (jonasasx) [reply upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:216]: I think, we must have this check. Otherwise `actual.GuestAccelerators` will be an empty slice instead of nil. It currupt back copatibility.

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 5 above)
Comment (jonasasx) [reply pkg/model/gcemodel/autoscalinggroup.go:171]: Why setting and unsetting slice is better than a preventive check?

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 2 above)
Comment (hakman) [reply pkg/model/gcemodel/autoscalinggroup.go:93]: In that case, the nil check is not needed.

File: pkg/apis/kops/instancegroup.go (hunk 1 above)
Comment (hakman) [reply pkg/apis/kops/instancegroup.go:372]: Not a big fan of this, but can live with it.

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go (hunk 3 above)
Comment (hakman) [reply upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:216]: Don't think I understand how it can corrupt backward compatibility.

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 5 above)
Comment (hakman) [reply pkg/model/gcemodel/autoscalinggroup.go:171]: tasks check actual vs model, when model is not `nil`.

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 2 above)
Comment (jonasasx) [reply pkg/model/gcemodel/autoscalinggroup.go]: fixed

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go (hunk 3 above)
Comment (jonasasx) [reply upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:216]: All InstanceTemplates created before this implementation will have a difference on applying this implementation:
```
- 
+ guestAccelerators: [],
```

Comment (hakman) [reply upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:216]: This is the kOps representation, not the internal one of GCP. `Find()` will return `[]AcceleratorConfig{}` wether `p.GuestAccelerators` is `nil` or empty slice.

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go (hunk 4 above)
Comment (hakman) [inline upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:351]: I think this is a little better.
```suggestion
	if len(e.GuestAccelerators) > 0 {
```

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 2 above)
Comment (hakman) [reply pkg/model/gcemodel/autoscalinggroup.go]: After reading a little, I think the correct way would be to set [Scheduling.OnHostMaintenance](https://pkg.go.dev/google.golang.org/api@v0.82.0/compute/v1#Scheduling.OnHostMaintenance).
`Preemptible ` seems to be the equivalent of Spot instances on AWS.

Comment (jonasasx) [reply pkg/model/gcemodel/autoscalinggroup.go]: Fixed. Moved to `InstanceTemplate.mapToGCE`

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go (hunk 3 above)
Comment (jonasasx) [reply upup/pkg/fi/cloudup/gcetasks/instancetemplate.go]: fixed

File: pkg/model/gcemodel/autoscalinggroup.go (hunk 5 above)
Comment (jonasasx) [reply pkg/model/gcemodel/autoscalinggroup.go]: fixed

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go [hunk 6]
```diff
@@ -235,7 +245,7 @@ func (e *InstanceTemplate) mapToGCE(project string, region string) (*compute.Ins
 	// TODO: This is similar to Instance...
 	var scheduling *compute.Scheduling
 
-	if fi.BoolValue(e.Preemptible) {
+	if fi.BoolValue(e.Preemptible) || len(e.GuestAccelerators) > 0 {
```
Comment (hakman) [inline upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:248]: This has nothing to do with preemption. Instead, add this after the current block:
```go
if len(e.GuestAccelerators) > 0 {
    // Instances with accelerators cannot be migrated.
    scheduling.OnHostMaintenance = "TERMINATE"
}

File: upup/pkg/fi/cloudup/gcetasks/instancetemplate.go [hunk 7]
```diff
@@ -335,13 +345,26 @@ func (e *InstanceTemplate) mapToGCE(project string, region string) (*compute.Ins
 		})
 	}
 
+	var accelerators []*compute.AcceleratorConfig
+	if len(e.GuestAccelerators) > 0 {
```
Comment (hakman) [inline upup/pkg/fi/cloudup/gcetasks/instancetemplate.go:349]: What happens when you remove accelerator from config? With this here it won't apply any change.

**Review by hakman — CHANGES_REQUESTED** [review]

File: upup/pkg/fi/cloudup/gcetasks/accelerator_config.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+func (a *AcceleratorConfig) GetDependencies(tasks map[string]fi.Task) []fi.Task {
+	return nil
+}
+
+func (_ *AcceleratorConfig) ShouldCreate(a, e, changes *AcceleratorConfig) (bool, error) {
+	if e.AcceleratorCount < 0 {
+		return false, fmt.Errorf("acceleratorCount must be positive or 0")
+	}
+	if e.AcceleratorCount < 0 {
```
Comment (hakman) [inline upup/pkg/fi/cloudup/gcetasks/accelerator_config.go:43]:
```suggestion
	if e.AcceleratorType == "" {
```

Comment (jonasasx) [reply upup/pkg/fi/cloudup/gcetasks/accelerator_config.go]: fixed

**Comment by hakman** [comment]
Please ignore the `pull-kops-e2e-k8s-gce` test failure, it is related to #13855.

**Review by hakman — APPROVED** [review]
